source/slang/ir-sccp.cpp

```diff
@@ -0,0 +1,933 @@
+
+        // We expect the caller to have filtered out functions with
+        // no bodies, so there should always be at least one basic block.
+        //
+        auto firstBlock = code->getFirstBlock();
+        SLANG_ASSERT(firstBlock);
+
+        // The entry block is always going to be executed when the
+        // function gets called, so we will process it right away.
+        //
+        cfgWorkList.Add(firstBlock);
+
+        // The parameters of the first block are our function parameters,
+        // and we want to operate on the assumption that they could have
+        // any value possible, so we will record that in our dictionary.
+        //
+        for( auto pp : firstBlock->getParams() )
+        {
+            setLatticeVal(pp, LatticeVal::getAny());
+        }
+
+        // Now we will iterate until both of our work lists go dry.
+        //
+        while(cfgWorkList.Count() || ssaWorkList.Count())
+        {
+            // Note: there is a design choice to be had here
+            // around whether we do `if if` or `while while`
+            // for these nested checks. The choice can affect
+            // how long things take to converge.
+
+            // We will start by processing any blocks that we
+            // have determined are potentially reachable.
+            //
+            while( cfgWorkList.Count() )
+            {
+                // We pop one block off of the work list.
+                //
+                auto block = cfgWorkList[0];
+                cfgWorkList.FastRemoveAt(0);
+
+                // We only want to process blocks that haven't
+                // already been marked as executed, so that we
+                // don't do redundant work.
+                //
+                if( !isMarkedAsExecuted(block) )
+                {
+                    // We should mark this new block as executed,
+                    // so we can ignore it if it ever ends up on
+                    // the work list again.
+                    //
+                    markAsExecuted(block);
+
+                    // If the block is potentially executed, then
+                    // that means the instructions in the block are too.
+                    // We will walk through the block and update our
+                    // guess at the value of each instruction, which
+                    // may in turn add other blocks/instructions to
+                    // the work lists.
+                    //
+                    for( auto inst : block->getChildren() )
+                    {
+                        updateValueForInst(inst);
+                    }
+                }
+            }
+
+            // Once we've cleared the work list of blocks, we
+            // will start looking at individual instructions that
+            // need to be updated.
+            //
+            while( ssaWorkList.Count() )
+            {
+                // We pop one instruction that needs an update.
```

Comment:
Same as before - if order isn't important (and doing a swap remove implies that), taking the last is typically faster... well assuming the calc to get the thing is effectively free cos small power of 2.